zipline/utils/events.py

```diff
@@ -372,23 +372,64 @@ def should_trigger(self, dt, env):
         return dt.date() not in env.early_closes
 
 
-class NthTradingDayOfWeek(StatelessRule):
+class TradingDayOfWeekRule(StatelessRule):
+    def __init__(self, n=0):
+        if not 0 <= abs(n) < MAX_WEEK_RANGE:
+            raise _out_of_range_error(MAX_WEEK_RANGE)
+
+        self.td_delta = n
+
+        self.next_date_start = None
+        self.next_date_end = None
+        self.next_midnight_timestamp = None
+
+        self.date_func = None
+
+    def calculate_start_and_end(self, dt, env):
+        next_trading_day = _coerce_datetime(
+            env.add_trading_days(
+                self.td_delta,
+                self.date_func(dt, env),
+            )
+        )
+
+        next_open, next_close = env.get_open_and_close(next_trading_day)
+        self.next_date_start = next_open
+        self.next_date_end = next_close
+        self.next_midnight_timestamp = \
+            pd.Timestamp(next_trading_day.date(), tz='UTC')
+
+    def should_trigger(self, dt, env):
+        if self.next_date_start is None:
+            # first time this method has been called.  calculate the midnight,
+            # open, and close of the next matching day.
+            self.calculate_start_and_end(dt, env)
+
+        # if the next matching day is in the past, calculate the next one.
+        if dt > self.next_date_end:
+            self.calculate_start_and_end(dt + datetime.timedelta(days=7),
+                                         env)
+
+        # if the given dt is within the next matching day, return true.
+        if self.next_date_start <= dt <= self.next_date_end or \
+                dt == self.next_midnight_timestamp:
+            return True
+
+        return False
+
+
+class NthTradingDayOfWeek(TradingDayOfWeekRule):
     """
     A rule that triggers on the nth trading day of the week.
     This is zero-indexed, n=0 is the first trading day of the week.
     """
     def __init__(self, n=0):
-        if not 0 <= n < MAX_WEEK_RANGE:
-            raise _out_of_range_error(MAX_WEEK_RANGE)
-        self.td_delta = n
+        super(NthTradingDayOfWeek, self).__init__(n)
 
-    def should_trigger(self, dt, env):
-        return _coerce_datetime(env.add_trading_days(
-            self.td_delta,
-            self.get_first_trading_day_of_week(dt, env),
-        )).date() == dt.date()
+        self.date_func = self.get_first_trading_day_of_week
```

Comment:
Should we make this a class-level function assignment, instead of setting it per instance?  Could also make `date_func` abstract in the base class.